File: vimana source code/crypto-lib/Java/cryptolib/src/main/java/io/vimana/cryptolib/dataformat/AEADMessage.java
package io.vimana.cryptolib.dataformat;

import io.vimana.cryptolib.vimryptoParams;

import java.nio.ByteBuffer;
import java.util.Arrays;

/**
 * Defines message format for AEAD with IV, plain authenticated
 * data and encrypted data.
 * specially formated data that includes: 
 *    IV  (12 bytes), (salt+explicit nounce)
 *    unencryped data lenght (4 bytes),
 *    ecnrypted data lenght (4 bytes)
 *    unencrypted data (variable len), 
 *    encrypted data in the rest of message including
 *    last 16 bytes (128 bits) of hmac 
 * @author alukin@gmail.com
 */
public class AEADMessage {
    /**
     * Maximal size of plain and encrypted parts in sum to prevent DoS attacks
     */
    public static final int MAX_MSG_SIZE = 65536;

    public byte[] aatext = new byte[0];
    public byte[] encrypted;

    private final int hmacSize;
    private final byte[] iv; //12 bytes = 4 of salt + 8 of nonce

    private final vimryptoParams cryptoParams;

    public AEADMessage(vimryptoParams cryptoParams) {
        this.cryptoParams = cryptoParams;
        this.hmacSize = cryptoParams.getGcmAuthTagLenBits() / 8; //128 bits
        this.iv = new byte[cryptoParams.getAesIvLen()]; //12 bytes, RFC 5288;  salt and explicit nounce
    }

    /**
     * Sets 8 bytes of implicit part on nounce that goes with message
     * @param en 8 bytes of explicit part of IV
     */
    public void setExplicitNounce(byte[] en){
        if(en.length!=8){
            throw new IllegalArgumentException("Nounce size must be exactly 8 bytes");
        }
        Arrays.fill(iv, (byte)0);
        System.arraycopy(en, 0, iv, 4, 8);       
    }

    
    public byte[] getExplicitNounce(){
        return Arrays.copyOfRange(iv, 4, 12);
    }

    public byte[] getIV(){
        return iv;
    }
    
    public void setIV(byte[] ivv){
       if(ivv.length != cryptoParams.getAesIvLen()){
            throw new IllegalArgumentException("Nounce size must be exactly 8 bytes");
        }
       System.arraycopy(ivv, 0, iv, 0, cryptoParams.getAesIvLen());
    }
    
    public byte[] getHMAC(){
      return Arrays.copyOfRange(encrypted, encrypted.length - hmacSize -1,encrypted.length-1);
    }
    
    public static AEADMessage fromBytes(byte[] message, vimryptoParams cryptoParams){
        AEADMessage res = new AEADMessage(cryptoParams);
        ByteBuffer bb = ByteBuffer.wrap(message);
        bb.get(res.iv);
        int txtlen = bb.getInt();
        int enclen = bb.getInt();
        //prevent overflow attack
        if(txtlen+enclen > MAX_MSG_SIZE){
            throw new IllegalArgumentException("Declared message size is too big: " + (txtlen + enclen));
        }
        res.aatext = new byte[txtlen];
        res.encrypted = new byte[enclen];
        bb.get(res.aatext);
        bb.get(res.encrypted);
        return res;
    }
    
    public byte[] toBytes(){
        int capacity = calcBytesSize();
        ByteBuffer bb = ByteBuffer.allocate(capacity);
        bb.put(iv);
        bb.putInt(aatext.length);
        bb.putInt(encrypted.length);
        bb.put(aatext);
        bb.put(encrypted); //hmac is 16 bytes tail of encrypted
        return bb.array();
    }

    public int calcBytesSize() {
        return iv.length + 4 + 4 + aatext.length + encrypted.length;
    }

}

File: vimana source code/crypto-lib/JS/cryptolib/src/app/dataformat/AEADMessage.js
import {byteArrayToHexString, utf8ToByteArray, hexStringToByteArray, stringToHexString} from 'bytey';


class AEADMessage {
    /**
     * Maximal size of plain and encrypted parts in sum to prevent DoS attacks
     */

    constructor() {

        this.MAX_MSG_SIZE = 65536;

        this.aatext;
        this.encrypted;
        this.hmacSize;
        this.iv;
        this.cryptoParams;
        this.explicitNounce;

    }

    setAEADParams(cryptoParams) {
        this.encrypted = cryptoParams.encrypted;
        this.aatext    = cryptoParams.aatext;
        this.hmacSize  = cryptoParams.hmacSize; //128 bits
        this.iv        = cryptoParams.iv;       //12 bytes, RFC 5288;  salt and explicit nounce
    }

    /**
     * Sets 8 bytes of implicit part on nounce that goes with message
     * @param en 8 bytes of explicit part of IV
     */
    setExplicitNounce(nonce){
        this.explicitNounce = nonce;
    }


    getExplicitNounce(){
        return this.explicitNounce;
    }

    getIV(){
        return this.iv;
    }

    setIV(iv){
        return this.iv;
    }

    getHMAC(){
        return Arrays.copyOfRange(encrypted, encrypted.length - hmacSize -1,encrypted.length-1);
    }

    // returns bytes
    fromBytes(message){

        let iv           = message.slice(0, 12);
        let salt         = message.slice(0,4);
        let nonce        = message.slice(4, 12);
        let aatextPos    = message.slice(12, 16)[3];
        let aatext       = message.slice(20, 20 + aatextPos);
        let cipherText   = message.slice(20 + aatextPos, message.length - 16);
        let authTag      = message.slice(message.length - 16, message.length);

        return {
            aatext,
            cipherText,
            iv,
            salt,
            nonce,
            authTag
        }

    }

    toBytes() {
        // let capacity = calcBytesSize(encrypted);
        let bb = [
            ...this.iv,            // 12 bytes
            ...[0,0,0],            // 4 bytes
            this.aatext.length,
            ...[0,0,0],            // 4 bytes
            this.encrypted.length,
            ...this.aatext,        //
            ...this.encrypted
        ];

        return byteArrayToHexString(bb);
    }

    calcBytesSize() {
        return iv.length + 4 + 4 + aatext.length + encrypted.length;



    }

}

export default AEADMessage;
